### PR TITLE
sbang: respect package permissive package permissions for sbang

### PR DESCRIPTION
**Pull request by spenpw**
When `sbang` is installed to the install directory, its permissions are hardcoded to 0o755. This change uses the permissions defined in `packages.yaml:packages:permissions:all` instead. This should allow multiple users to install packages with different versions of Spack to the same install directory cleanly if they are members of the same group and that group is specified in `packages:all:permissions:write`

### CODE REVIEW

**Comment by spenpw** [comment]
@tgamblin Could you or someone else take a look at this? Thanks

**Comment by tgamblin** [comment]
Thanks @pwablito!  We'll get to this soon.

**Comment by spenpw** [comment]
@tgamblin today I noticed that some json database files are also created as user:user rather than user:group. Would it be appropriate to make a change similar to this PR adjusting that?

**Review by RikkiButler20 — DISMISSED** [review]

**Comment by tgamblin** [comment]
@pwablito yep that would be useful too.

**Comment by spenpw** [comment]
> @pwablito yep that would be useful too.

Would you like me to add that here and make this PR more of a broad "move away from legacy permissions" change? Or open a new one?

**Comment by tgamblin** [comment]
> Would you like me to add that here and make this PR more of a broad "move away from legacy permissions" change? Or open a new one?

New one is better but I do have one change request incoming on this one -- if you could do that first and open a separate one for the other thing it'd be great!

File: lib/spack/spack/hooks/sbang.py
Comment (tgamblin) [inline lib/spack/spack/hooks/sbang.py]: Downside here is that if we remove sbang first thing, then if anything goes wrong below we lose our `sbang`.  Also, if there are lots of jobs on the machine, they can potentially *see* `sbang` blink in and out of existence with this approach.  You'll see scripts fail mysteriously saying there is no `sbang` if they happen to have a race with the `sbang` installation logic here.

Can you modify this to copy and chmod a file, `.sbang.tmp`, alongside `sbang`, then move it into place *last*?

File: lib/spack/spack/hooks/sbang.py
Comment (tgamblin) [inline lib/spack/spack/hooks/sbang.py]: Something like:

```suggestion
    sbang_tmp_path = os.path.join(
        os.path.dirname(sbang_path),
        ".%s.tmp" % os.path.basename(sbang_path),
    )
    shutil.copy(spack.paths.sbang_script, sbang_tmp_path)
```

File: lib/spack/spack/hooks/sbang.py
Comment (tgamblin) [inline lib/spack/spack/hooks/sbang.py]:
```suggestion
    os.chmod(sbang_tmp_path, config_mode)
    if group_name:
        os.chown(
            sbang_tmp_path,
            os.stat(sbang_tmp_path).st_uid,
            grp.getgrnam(group_name).gr_gid
        )
    
    # Finally, move the new `sbang` into place atomically
    os.rename(sbang_tmp_path, sbang_path)
```

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/hooks/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/hooks/sbang.py:196]: IIRC group members must still have at least execute permissions on the `sbang_bin_dir` in order to traverse/access contents of the directory.  I recommend retaining the permissions call that was at line 154, i.e.:

```suggestion
    fs.set_install_permissions(sbang_bin_dir)
    
```

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/test/sbang.py]: Assuming `sbang_bin_dir` permissions update is retained, this check should also be restored.

**Review by tldahlgren — CHANGES_REQUESTED** [review]

**Comment by tldahlgren** [comment]
Timing!  Looks like I should've refreshed to pick up Todd's recommendations before I submitted mine.

File: lib/spack/spack/hooks/sbang.py
Comment (haampie) [reply lib/spack/spack/hooks/sbang.py:156]: I guess the idea is that you have to delete the file because you may not have permission to chmod it?

File: lib/spack/spack/hooks/sbang.py
Comment (spenpw) [reply lib/spack/spack/hooks/sbang.py:196]: This would set the directory permissions to 0o755, which doesn't allow other members of your the group to write files (like `sbang`). That's why I got rid of it.

File: lib/spack/spack/test/sbang.py
Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: Here we could instead check the config to see if 0o755 or 0o775 should be set (see if a group is specified), then check if it as actually set correctly.

File: lib/spack/spack/hooks/sbang.py
Comment (spenpw) [reply lib/spack/spack/hooks/sbang.py]: Great suggestion! I tested out your suggested changes, they worked great for my use case and didn't break anything, so I accepted them.

Comment (tgamblin) [reply lib/spack/spack/hooks/sbang.py]: @pwablito: the initial remove (this code) need to be deleted as well

Comment (spenpw) [reply lib/spack/spack/hooks/sbang.py]: Ah, forgot to do that. Fixed in 3c31801

**Comment by tgamblin** [comment]
So the only unresolved things here are @tldahlgren's concerns.  Right now it looks like the sbang directory is no longer chmod'd (only chown'd) so we're relying on the user's umask to set permissions.  I think the permissions need to mirror the settings in config -- i.e., writable by group if `all` packages are supposed to be writable by group, etc.

File: lib/spack/spack/hooks/sbang.py
Comment (tldahlgren) [reply lib/spack/spack/hooks/sbang.py:157]: I thought line 180 sets it.  In my testing, group and world ha no access to the directory whatsoever even when I set the package permissions to give write access.

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py]: Whatever makes sense.

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py]: My point was that it would be good to make sure the directory has suitable permissions.

File: lib/spack/spack/hooks/sbang.py
Comment (spenpw) [reply lib/spack/spack/hooks/sbang.py:196]: @tldahlgren I just updated in 0e3f4f0 - is that what you had in mind?

File: lib/spack/spack/test/sbang.py
Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: @tldahlgren I added this test in bb9299d

File: lib/spack/spack/hooks/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/hooks/sbang.py]: I was thinking more like:

```suggestion
    if group_name:
        os.chmod(sbang_bin_dir, config_mode)   # Use package directory permissions
    else:
        fs.set_install_permissions(sbang_bin_dir)
```

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/test/sbang.py]: Why not check the directory permissions at ~line 219? 

Also, you could keep the directory perms check with `check_sbang_installation` by restoring line 194, grabbing the expected mode for the directory using `config_mode = spack.package_prefs.get_package_dir_permissions(spack.spec.Spec("all"))` and using a single assert `assert (status.st_mode & 0o777) == config_mode` for the directory.

**Review by tldahlgren — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/sbang.py
Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: I am getting config_mode = 0o700 when I do it this way. Not sure why.

Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: That's causing it to fail the tests.

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py:200]: Those pesky `umasks`.  And there isn't a good/safe way to check the current value without changing it, which can be problematic.

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py:200]: Do the checks you removed at lines 200-201 fail for you?

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py]: Manually checking the sbang changes as of `276a1a7b141f587657d496de077e3bf8dacab4df ` works for me but I experienced the same issues with tests.

I have a fix for the tests but don't have permissions to push them to your branch.  

Basically I am suggesting adding the following import:

```
import spack.util.spack_yaml as syaml
```

*AND* replacing the current contents of lines 192 on of `lib/spack/spack/test/sbang.py` with the code below.  The fixture should ensure consistent permissions.  (I'm not specifying a group here because we can't guarantee a benign group across all systems.)  I'm curious if this will work for you *and* if it will pass the PR checks.

```
@pytest.fixture(scope='function')
def configure_world_perms():
    conf = syaml.load_config("""\
all:
  permissions:
    read: world
    write: world
""")
    spack.config.set('packages', conf, scope='user')

    yield


def check_sbang_installation():
    sbang_path = sbang.sbang_install_path()
    sbang_bin_dir = os.path.dirname(sbang_path)
    assert sbang_path.startswith(spack.store.store.unpadded_root)

    assert os.path.exists(sbang_path)
    assert fs.is_exe(sbang_path)

    status = os.stat(sbang_bin_dir)
    mode = (status.st_mode & 0o777)
    assert mode == 0o755, 'Unexpected {0}'.format(oct(mode))

    status = os.stat(sbang_path)
    mode = (status.st_mode & 0o777)
    assert mode == 0o777, 'Unexpected {0}'.format(oct(mode))


def test_install_sbang(install_mockery, configure_world_perms):
    sbang_path = sbang.sbang_install_path()
    sbang_bin_dir = os.path.dirname(sbang_path)

    assert sbang_path.startswith(spack.store.store.unpadded_root)
    assert not os.path.exists(sbang_bin_dir)

    sbang.install_sbang()
    check_sbang_installation()

    # put an invalid file in for sbang
    fs.mkdirp(sbang_bin_dir)
    with open(sbang_path, "w") as f:
        f.write("foo")

    sbang.install_sbang()
    check_sbang_installation()

    # install again and make sure sbang is still fine
    sbang.install_sbang()
    check_sbang_installation()
```

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py:200]: The following import is also needed for the above to work:

```
import spack.util.spack_yaml as syaml
```

Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: You're right, that did fix the tests (and I added your changes in 527338b). However, that only ever tests a situation where the group permissions are not set. Should we add another test case to verify the installation of `sbang` with group write permissions?

Comment (tldahlgren) [reply lib/spack/spack/test/sbang.py]: Configuring to 'world', IIRC, automatically sets group permissions.

AFAIK The only way to make such a test portable would be to monkey patch the system call(s) for a test group.  I'm not sure it's worth the extra effort.  But it's up to you.

Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: Not too worried about it, this seems good to me.

**Review by tldahlgren — APPROVED** [review]
LGTM .. except I missed the style issue. 😦

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/test/sbang.py]: The following will fix the style error:

```suggestion
import spack.util.spack_yaml as syaml
from spack.util.executable import which
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]

**Review by tldahlgren — DISMISSED** [review]

**Comment by tldahlgren** [comment]
@pwablito Looks like world writable is an issue for some tests.  

🤦 It hadn't dawned on me before but one possibility is to change `configure_world_perms` to customize the group to that of the account running the program and see if that will work.  You could then adjust the tests accordingly.

**Comment by spenpw** [comment]
@tldahlgren I just pushed 3b1077f which tests for both user and group write cases, and users the group of the running process for the group write permissions.

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/test/sbang.py]:
```suggestion
    group: {0}
```

File: lib/spack/spack/test/sbang.py
Comment (tldahlgren) [inline lib/spack/spack/test/sbang.py]: Since this duplicates lines 268-287, I recommend creating a method, such as `run_test_install_sbang(group)` that you call from `test_install_group_sbang` and `test_install_user_sbang` with the appropriate group value to simplify the code.

**Review by tldahlgren — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/sbang.py
Comment (spenpw) [reply lib/spack/spack/test/sbang.py]: Taken care of in a616206

**Comment by spenpw** [comment]
@tldahlgren I am not quite sure why the pipelines aren't running

**Review by tldahlgren — DISMISSED** [review]

**Comment by tldahlgren** [comment]
Can you rebase to develop and resolve any conflicts?  For some reason this PR is flagged with conflicts.

**Comment by spenpw** [comment]
> Can you rebase to develop and resolve any conflicts? For some reason this PR is flagged with conflicts.

Just did :)

**Comment by spenpw** [comment]
@tldahlgren I just realized that things got jumbled in the merge conflict resolution and I had accidentally added another round of `sbang` installation and checking to the tests, so I fixed that... So now we get to wait for another round of pipelines to finish.

**Comment by spenpw** [comment]
@tldahlgren Everything here seems to be good to go except for the CI pipeline for some reason.

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Comment by spenpw** [comment]
@tldahlgren Looks like the pipeline will fail with the same issue as last time. Not sure what exactly the problem is but it seems unrelated to this PR.

**Comment by spenpw** [comment]
@tldahlgren A rebase fixed the problem.

**Review by tldahlgren — APPROVED** [review]

**Comment by tldahlgren** [comment]
@RikkiButler20 This PR can't be merged until you approve it.

**Review by RikkiButler20 — APPROVED** [review]

**Comment by tldahlgren** [comment]
@tgamblin Looks like your review is blocking the merge of this PR.

**Comment by spenpw** [comment]
FYI I am graduating this week so I will get kicked off the organization that owns this fork. If @tgamblin or @tldahlgren or someone else needs anything else on here I may have to open a new PR from my own fork. Just lmk :)

**Comment by tldahlgren** [comment]
> FYI I am graduating this week so I will get kicked off the organization that owns this fork. If @tgamblin or @tldahlgren or someone else needs anything else on here I may have to open a new PR from my own fork. Just lmk :)

Congratulations on your graduation!

**Review by tgamblin — APPROVED** [review]

File: lib/spack/spack/hooks/sbang.py
Comment (haampie) [inline lib/spack/spack/hooks/sbang.py:221]: This leads to race conditions. Why do you need a temporary file here?

**Review by haampie — COMMENTED** [review]
This change introduces race conditions in `spack install`.

File: lib/spack/spack/hooks/sbang.py
Comment (haampie) [reply lib/spack/spack/hooks/sbang.py]: Unfortunately this is what leads to races.

File: lib/spack/spack/hooks/sbang.py
Comment (tgamblin) [reply lib/spack/spack/hooks/sbang.py:221]: you have to get the new `sbang` next to the destination before moving it, or you may end up with a non-atomic move into place.  This solves the issue that the real `sbang` can't be invoked in a partial state of copying, but now there's a race on the tempfile.  We either need to use a lock here, or we need to use the tmpfile API to generate a race-free name for the copy.

Comment (haampie) [reply lib/spack/spack/hooks/sbang.py:221]: Do we really care about that aspect though? There's already a guard that does a file compare, so we install sbang approximately 1 times ever, and at that point there's no race because there is no sbang.

Comment (haampie) [reply lib/spack/spack/hooks/sbang.py:221]: See https://github.com/spack/spack/pull/33549 for a fix.

**Pull request merged**
